[JIT] during freezing, cast optional bias to half if weight is half

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77295

On GPU, conv errors if not all its inputs have the same dtype.

In the case of autocasting during freezing, what we see is:
1) inputs to conv are casted to half
2) inputs to batchnorm are not casted, so many are still floats
3) we try to fold conv + batchnorm, by finding different weight and bias such that conv(input, new_weight, new_bias) is equivalent to the original conv -> batchnorm.

If conv previously had an optional bias, then during freezing we will temporarily create a zero-valued bias as a placeholder for conv_bias. We want to construct it to have the same dtype as the weight input to conv, to avoid errors on GPU.

Fixes #76578

davidberard98 committed:
[JIT] during freezing, cast optional bias to half if weight is half

On GPU, conv errors if not all its inputs have the same dtype.

In the case of autocasting during freezing, what we see is:
1) inputs to conv are casted to half
2) inputs to batchnorm are not casted, so many are still floats
3) we try to fold conv + batchnorm, by finding different weight and bias such that conv(input, new_weight, new_bias) is equivalent to the original conv -> batchnorm.

If conv previously had an optional bias, then during freezing we will temporarily create a zero-valued bias as a placeholder for conv_bias. We want to construct it to have the same dtype as the weight input to conv, to avoid errors on GPU.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] during freezing, cast optional bias to half if weight is half"


On GPU, conv errors if not all its inputs have the same dtype.

In the case of autocasting during freezing, what we see is:
1) inputs to conv are casted to half
2) inputs to batchnorm are not casted, so many are still floats
3) we try to fold conv + batchnorm, by finding different weight and bias such that conv(input, new_weight, new_bias) is equivalent to the original conv -> batchnorm.

If conv previously had an optional bias, then during freezing we will temporarily create a zero-valued bias as a placeholder for conv_bias. We want to construct it to have the same dtype as the weight input to conv, to avoid errors on GPU.

Fixes #76578

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] during freezing, cast optional bias to half if weight is half"


On GPU, conv errors if not all its inputs have the same dtype.

In the case of autocasting during freezing, what we see is:
1) inputs to conv are casted to half
2) inputs to batchnorm are not casted, so many are still floats
3) we try to fold conv + batchnorm, by finding different weight and bias such that conv(input, new_weight, new_bias) is equivalent to the original conv -> batchnorm.

If conv previously had an optional bias, then during freezing we will temporarily create a zero-valued bias as a placeholder for conv_bias. We want to construct it to have the same dtype as the weight input to conv, to avoid errors on GPU.

Fixes #76578

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] during freezing, cast optional bias to half if weight is half"


On GPU, conv errors if not all its inputs have the same dtype.

In the case of autocasting during freezing, what we see is:
1) inputs to conv are casted to half
2) inputs to batchnorm are not casted, so many are still floats
3) we try to fold conv + batchnorm, by finding different weight and bias such that conv(input, new_weight, new_bias) is equivalent to the original conv -> batchnorm.

If conv previously had an optional bias, then during freezing we will temporarily create a zero-valued bias as a placeholder for conv_bias. We want to construct it to have the same dtype as the weight input to conv, to avoid errors on GPU.

Fixes #76578

[ghstack-poisoned]

davidberard98 commented:
cc @ngimel this is the test where we're seeing the cuda caching allocator failures:
```
$ python test/test_jit.py -k test_conv_bn_folding_autocast
... caching allocator leaks 2560 ...
$ PYTORCH_NO_CUDA_MEMORY_CACHING=1 python test/test_jit.py -k test_conv_bn_folding_autocast
... succeeds ...
```

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1678,6 +1678,30 @@ def make_prediction(self, x):
+        mod_eager = ConvBN(3, 32, kernel_size=3, stride=2).cuda().eval()
+        scripted_mod = torch.jit.script(mod_eager)
+        scripted_mod = torch.jit.freeze(scripted_mod)
+        FileCheck().check("conv").check_not("aten::batch_norm").run(scripted_mod.graph)
```
Could you add a check that the `conv` weight here is `float` ?

eellison replied:
sorry, `conv` weight is `half`

eellison commented on torch/csrc/jit/passes/frozen_conv_folding.cpp:
```diff
@@ -79,7 +79,20 @@ void FoldFrozenConvBatchnorm(Block* b) {
+        at::ScalarType bias_dtype = bn_rm.scalar_type();
+        at::ScalarType weight_dtype = conv_w.scalar_type();
+        at::DeviceType weight_device = conv_w.device().type();
+        if (weight_device == at::kCUDA &&
```
is this any reason to not just always put `conv_b` as the same dtype as `weight_dtype` ? for cpu as well ?

eellison reviewed (approved):
LGTM! Just a couple of small questions

davidberard98 commented on torch/csrc/jit/passes/frozen_conv_folding.cpp:
```diff
@@ -79,7 +79,20 @@ void FoldFrozenConvBatchnorm(Block* b) {
+        at::ScalarType bias_dtype = bn_rm.scalar_type();
+        at::ScalarType weight_dtype = conv_w.scalar_type();
+        at::DeviceType weight_device = conv_w.device().type();
+        if (weight_device == at::kCUDA &&
```
on CPU this won't error out, so maybe there's some use case where users want bias to be a float.

^ but if you think that's not really an issue, it's probably fine to always cast

davidberard98 committed:
Update on "[JIT] during freezing, cast optional bias to half if weight is half"


On GPU, conv errors if not all its inputs have the same dtype.

In the case of autocasting during freezing, what we see is:
1) inputs to conv are casted to half
2) inputs to batchnorm are not casted, so many are still floats
3) we try to fold conv + batchnorm, by finding different weight and bias such that conv(input, new_weight, new_bias) is equivalent to the original conv -> batchnorm.

If conv previously had an optional bias, then during freezing we will temporarily create a zero-valued bias as a placeholder for conv_bias. We want to construct it to have the same dtype as the weight input to conv, to avoid errors on GPU.

Fixes #76578

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge this please

malfet commented:
@pytorchmergebot revert this, as it causes failures in newly added tests, which were occluded in PR CI by pre-existing failures failures, see https://hud.pytorch.org/pytorch/pytorch/commit/2547be51354145b18a2430bd3e11ffae6c736c6e
